test/rateLimiters/fixedWindow.test.ts

```diff
@@ -0,0 +1,128 @@
+const RedisMock = require('ioredis-mock');
+
+const CAPACITY = 10;
+const WINDOW_SIZE = 6000;
+
+let limiter: FixedWindow;
+let client: ioredis.Redis;
+let timestamp: number;
+const user1 = '1';
+const user2 = '2';
+const user3 = '3';
+
+async function getWindowFromClient(redisClient: ioredis.Redis, uuid: string): Promise<RedisWindow> {
+    const res = await redisClient.get(uuid);
+    // if no uuid is found, return -1 for tokens and timestamp, which are both impossible
+    if (res === null) return { currentTokens: -1, fixedWindowStart: -1 };
+    return JSON.parse(res);
+}
+
+async function setTokenCountInClient(
+    redisClient: ioredis.Redis,
+    uuid: string,
+    currentTokens: number,
+    fixedWindowStart: number
+) {
+    const value: RedisWindow = { currentTokens, fixedWindowStart };
+    await redisClient.set(uuid, JSON.stringify(value));
+}
+describe('Test FixedWindow Rate Limiter', () => {
+    beforeEach(async () => {
+        client = new RedisMock();
+        limiter = new FixedWindow(CAPACITY, WINDOW_SIZE, client);
+        timestamp = new Date().valueOf();
+    });
+    describe('FixedWindow returns correct number of tokens and updates redis store as expected', () => {
+        describe('after an ALLOWED request...', () => {
+            afterEach(() => {
+                client.flushall();
+            });
+            test('current time window has no token initially', async () => {
+                // zero token used in this time window
+                const withdraw5 = 5;
+                expect((await limiter.processRequest(user1, timestamp, withdraw5)).tokens).toBe(
+                    CAPACITY - withdraw5
+                );
+                const tokenCountFull = await getWindowFromClient(client, user1);
+                expect(tokenCountFull.currentTokens).toBe(5);
+            });
+            test('reached 40% capacity in current time window and still can pass request', async () => {
+                const initial = 5;
+                const partialWithdraw = 2;
+                expect(
+                    (await limiter.processRequest(user2, timestamp, initial + partialWithdraw))
+                        .tokens
+                ).toBe(CAPACITY - initial - partialWithdraw);
+
+                const tokenCountPartial = await getWindowFromClient(client, user2);
+                expect(tokenCountPartial.currentTokens).toBe(initial + partialWithdraw);
+            });
+
+            test('window is partially full and request has no leftover tokens', async () => {
+                const initial = 6;
+                const partialWithdraw = 5;
+                await setTokenCountInClient(client, user2, initial, timestamp);
+                expect(
+                    (await limiter.processRequest(user2, timestamp, partialWithdraw)).success
+                ).toBe(false);
+                expect(
+                    (await limiter.processRequest(user2, timestamp, partialWithdraw)).tokens
+                ).toBe(0);
```

Comment:
I would expect tokens to be `4` here because:
1. your setting `6` in the cache (4 available)
2. trying to request 5 more (fail)
3. but the there should still be 4 available afterwards

Reply:
Will change in the revised version.